From de764ac86dc6ecb311b896f770134755dc022bac Mon Sep 17 00:00:00 2001
From: Cat73 <1901803382@qq.com>
Date: Tue, 6 Dec 2016 15:13:03 +0800
Subject: [PATCH] Cat73 -> Improve Minecraft Hopper Performance
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

在漏斗传输物品的时候不再调用目标容器的 update 方法
简单看了看似乎这个 update 在传输物品的时候并没有什么卵用。。。
1. 更新方块的损害值。
2. 更新 Chunk 里是否有过方块更新的标记为 true。
   然而只有很低的概率影响 MC 每隔一段时间的自动保存，卸载 Chunk 时一定会保存。
   AutoSaveWorld 也是无视这个值直接保存的，服务器崩溃时 AutoSaveWorld 会先保存一次再关闭服务器。
   而且服务器当前已经在 bukkit.yml 里禁用了这种自动保存。。。
3. 进行 Physics 计算。
Link: https://github.com/PaperMC/Paper/blob/ebf1094ea5dea6f3c8ad0ed5a1344089cb7ecfc7/Spigot-Server-Patches/0131-Improve-Minecraft-Hopper-Performance.patch

diff --git a/src/main/java/net/minecraft/server/TileEntity.java b/src/main/java/net/minecraft/server/TileEntity.java
index 8d54eb36..6cdf1442 100644
--- a/src/main/java/net/minecraft/server/TileEntity.java
+++ b/src/main/java/net/minecraft/server/TileEntity.java
@@ -108,8 +108,10 @@ public abstract class TileEntity {
         return this.g;
     }
 
+    static boolean IGNORE_TILE_UPDATES = false; // Paper
     public void update() {
         if (this.world != null) {
+            if (IGNORE_TILE_UPDATES) return; // Paper
             IBlockData iblockdata = this.world.getType(this.position);
 
             this.g = iblockdata.getBlock().toLegacyData(iblockdata);
diff --git a/src/main/java/net/minecraft/server/TileEntityHopper.java b/src/main/java/net/minecraft/server/TileEntityHopper.java
index f12bc709..c86bddb1 100644
--- a/src/main/java/net/minecraft/server/TileEntityHopper.java
+++ b/src/main/java/net/minecraft/server/TileEntityHopper.java
@@ -455,7 +455,9 @@ public class TileEntityHopper extends TileEntityLootable implements IHopper, ITi
             boolean flag1 = iinventory1.w_();
 
             if (itemstack1.isEmpty()) {
+                IGNORE_TILE_UPDATES = true; // Paper
                 iinventory1.setItem(i, itemstack);
+                IGNORE_TILE_UPDATES = false; // Paper
                 itemstack = ItemStack.a;
                 flag = true;
             } else if (a(itemstack1, itemstack)) {
-- 
2.11.0.windows.1

